src/app-layer-htp-range.c

```diff
@@ -0,0 +1,570 @@
+        curf->container = c;
+        curf->fileOwning = true;
+        // If we see 2 duplicate range requests with the same range,
+        // the first one takes the ownership with fileOwned
+        // protected by the lock from caller HTPFileOpenWithRange
+        c->fileOwned = true;
+        return curf;
+    } else if (start < c->lastsize && c->lastsize - start >= buflen) {
+        // only overlap
+        // redundant to be explicit that this block is independent
+        curf->toskip = buflen;
+        return curf;
+    } else if (start < c->lastsize && c->lastsize - start < buflen && !c->fileOwned) {
+        // some overlap, then some data to append to the file
+        curf->toskip = c->lastsize - start;
+        curf->fileOwning = true;
+        c->fileOwned = true;
+        curf->container = c;
+        return curf;
+    }
+    // Because we are not in the previous cases, we will store the data for later use
+
+    // block/range to be inserted in ordered linked list
+    if (!(THASH_CHECK_MEMCAP(ContainerUrlRangeList.ht, buflen))) {
+        // skips this range
+        curf->toskip = buflen;
+        return curf;
+    }
+    curf->container = c;
+
+    HttpRangeContainerBuffer *range = SCCalloc(1, sizeof(HttpRangeContainerBuffer));
+    if (range == NULL) {
+        c->error = true;
+        SCFree(curf);
+        return NULL;
+    }
+
+    (void)SC_ATOMIC_ADD(ContainerUrlRangeList.ht->memuse, buflen);
+    range->buffer = SCMalloc(buflen);
+    if (range->buffer == NULL) {
+        c->error = true;
+        SCFree(curf);
+        SCFree(range);
+        (void)SC_ATOMIC_SUB(ContainerUrlRangeList.ht->memuse, buflen);
+        return NULL;
+    }
+    range->buflen = buflen;
+    range->start = start;
+    range->offset = 0;
+    range->gap = 0;
+    curf->current = range;
+    return curf;
+}
+
+HttpRangeContainerBlock *HttpRangeOpenFile(HttpRangeContainerFile *c, uint64_t start, uint64_t end,
+        uint64_t total, const StreamingBufferConfig *sbcfg, const uint8_t *name, uint16_t name_len,
+        uint16_t flags, const uint8_t *data, uint32_t len)
+{
+    HttpRangeContainerBlock *r =
+            HttpRangeOpenFileAux(c, start, end, total, sbcfg, name, name_len, flags);
+    if (HttpRangeAppendData(r, data, len) < 0) {
+        SCLogDebug("Failed to append data while openeing");
+    }
+    return r;
+}
+
+int HttpRangeAppendData(HttpRangeContainerBlock *c, const uint8_t *data, uint32_t len)
+{
+    if (len == 0) {
+        return 0;
+    }
+    // first check if we need to skip all bytes
+    if (c->toskip >= len) {
+        c->toskip -= len;
+        return 0;
+    }
+    // then if we need to skip only some bytes
+    if (c->toskip > 0) {
+        int r;
+        if (c->fileOwning) {
+            DEBUG_VALIDATE_BUG_ON(c->container->files == NULL);
+            if (data == NULL) {
+                // gap overlaping already known data
+                r = FileAppendData(c->container->files, NULL, len - c->toskip);
```

Comment:
it looks like this is accessed w/o a lock in the `HTPFileStoreChunk` path. It may work due to `c->fileOwning`, but I'm not happy with this setup. I don't want to see any access of a shared data structure w/o holding a lock.

Like I wrote in another comment, I thought the idea was to move `c->container->files` into `HttpRangeContainerBlock` for the flow handling the in-order chunk. That way we can append to it w/o needing to access the shared data at `c->container`

Reply:
Reworking with the `FileContainer` pointer from comment above